### PR TITLE
Document OpsHandler protocol

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #117862
* #117859
* __->__ #117790

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/_inductor/virtualized.py [hunk 1]
```diff
@@ -189,6 +206,451 @@ def inner(*args):
...
+    def constant(self, value: Union[bool, float, int], dtype: torch.dtype) -> T:
+        """Produces a scalar constant of type dtype."""
+        ...
+
+    # TODO: not sure about sympy.Expr here
+    def load_seed(self, name: str, offset: sympy.Expr):
+        """Computes inductor_prims.lookup_seed."""
+        ...
+
+    def rand(self, seed: T, offset: T) -> T:
```
Comment (ezyang) [inline torch/_inductor/virtualized.py:253]: Hmm... actually maybe offset is a sympy.Expr?

File: torch/_inductor/codegen/common.py [hunk 2]
```diff
@@ -1052,26 +1059,40 @@ def indirect_load(self, name: str, index: sympy.Expr):
...
+    ):
         raise NotImplementedError()
 
-    def reduction(self, dtype, src_dtype, reduction_type, value):
+    def reduction(
+        self,
+        dtype: torch.dtype,
+        src_dtype: torch.dtype,
+        reduction_type: ReductionType,
+        value: CSEVariable,
```
Comment (peterbell10) [inline torch/_inductor/codegen/common.py:1075]: For `reduction_type="welford_combine"` value is a tuple `CSEVariable`
```suggestion
        value: Union[CSEVariable, Tuple[CSEVariable, ...]],
```

File: torch/_inductor/codegen/common.py [hunk 3]
```diff
@@ -1181,7 +1205,9 @@ def load(name: str, index: sympy.Expr):
                 return self.load(name, index)
 
             @staticmethod
-            def store(name, index, value, mode=None):
+            def store(
+                name: str, index: sympy.Expr, value: CSEVariable, mode: StoreMode = None
+            ) -> CSEVariable:
```
Comment (peterbell10) [inline torch/_inductor/codegen/common.py:1210]: I don't think store does return a value during codegen, even though `MockHandler` does.

File: torch/_inductor/codegen/common.py [hunk 4]
```diff
@@ -1052,26 +1059,40 @@ def indirect_load(self, name: str, index: sympy.Expr):
...
+    ) -> Union[CSEVariable, Tuple[CSEVariable, ...]]:
         raise NotImplementedError()
 
-    def scan(self, dtype, combine_fn, value, init):
+    def scan(
+        self,
+        dtype: torch.dtype,
+        combine_fn: Callable[[CSEVariable, CSEVariable], CSEVariable],
+        value: CSEVariable,
+        init: CSEVariable,
```
Comment (peterbell10) [inline torch/_inductor/codegen/common.py:1084]: I believe `init` is a number at the moment.

File: torch/_inductor/virtualized.py (hunk 1 above)
Comment (peterbell10) [reply torch/_inductor/virtualized.py:273]: Looks like the lowerings call `ops.index_expr` before passing the expression to `ops.rand`, so this is correct as-is.

File: torch/_inductor/virtualized.py [hunk 5]
```diff
@@ -189,6 +219,460 @@ def inner(*args):
...
+
+    def where(self, condition: T, input: T, other: T) -> T:
+        """
+        Computes torch.where: when condition is true, return input; otherwise return other.
+        """
+        ...
+
+    def index_expr(self, expr: sympy.Expr, dtype: torch.dtype) -> T:
+        """
+        Converts a sympy expression into a scalar of type expr.  expr is typically
```
Comment (peterbell10) [inline torch/_inductor/virtualized.py:306]: Should this be
```suggestion
        Converts a sympy expression into a scalar of type dtype.  expr is typically
```

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**